blog: add predicted-latency based scheduling for LLMs (#208)
llm-d/llm-d.github.io: blog/2026-03-13_predicted-latency-based-scheduling-for-llms.md

| **Input Tokens In Flight** | Input tokens dispatched but not yet prefilled, plus input tokens already prefilled but still occupying KV cache (request not complete) | Captures both incoming prefill pressure and lingering memory footprint -> higher TTFT; helps the model anticipate load before it hits the server |

#### Training Data

Performance characteristics in LLM serving are batch-dependent and shift too quickly for long-term historical averages to remain meaningful. To stay aligned with real traffic, the model:

- Collects the most recent samples using a **sliding window**
- Stratifies them into coarse **buckets** (KV cache % in steps of 10, prefix hit rate in steps of 0.25, etc.)
- Continuously retrains on this stratified dataset

Bucketing with a sliding window is important because it maintains samples from regimes that aren't showing up in the latest traffic. Without bucketing, a single global sliding window would let the newest data overwrite everything. For example, if current traffic sits around 60% KV cache utilization, older samples from 30% KV cache would eventually disappear -- and the model would forget how to predict in that regime.

### Request Scheduling

> The latency predictor runs as a sidecar to the [Inference Gateway Endpoint Picker](https://gateway-api-inference-extension.sigs.k8s.io/) (EPP), hosting both training and prediction servers:

Reviewer: We may call this a sidecar but that is not the technical definition of the term - sidecars run inside the same pod along the main container, where as we deploy prediction and training servers as their own deployments:

```suggestion
The latency predictor is comprised of deployments of training and prediction servers, as well as a plugin to the [Inference Gateway Endpoint Picker](https://gateway-api-inference-extension.sigs.k8s.io/) (EPP):
```

Reviewer: Ah I was looking at the https://github.com/kubernetes-sigs/gateway-api-inference-extension/blob/main/latencypredictor/manifests/dual-server-deployment.yaml which was structured as per my comment, but as you point out the most important thing is the chart

The default threshold of 0.80 comes from production observation: prefix cache scores follow a bimodal distribution, roughly half of request-pod pairs have very high cache match (\>0.80) and half have low match (\<0.80). This reflects how prefix caching works in practice: in multi-turn conversations, a pod either has the conversation history cached from prior turns or it doesn't. Partial matches from unrelated conversations contribute very little because caching is block-based. The 0.80 threshold cleanly separates these two populations, so the affinity gate routes to pods that genuinely have your conversation cached rather than pods with incidental partial matches.

<div style={{textAlign: 'center', margin: '20px 0'}}>
<img src="/img/blogs/predicted-latency/prefix_hit_ratio_dist.png" alt="Prefix cache hit ratio distribution showing bimodal pattern" style={{width: '85%', height: 'auto'}} />
<p style={{fontSize: '0.9em', marginTop: '8px'}}><em>A typical production prefix hit ratio distribution observed in internal workloads, showing the bimodal pattern that motivates the 0.80 affinity threshold.</em></p>
</div>

## Benchmark Scenario Comparison

> The table below contrasts five scenarios, ranging from cache-friendly (high prefix-sharing) to cache-intensive scenarios. We used the [**inference-perf**](https://github.com/kubernetes-sigs/inference-perf/pull/301) library to enable shared prefix benchmarking configurations with multi-chat support. See [Appendix](#appendix) for a complete analysis of the workloads.

Reviewer: redundant + multi-turn fix
```suggestion
The table below contrasts five scenarios, ranging from cache-friendly (high prefix-sharing) to cache-intensive. We used the [**inference-perf**](https://github.com/kubernetes-sigs/inference-perf/pull/301) library to enable shared prefix benchmarking configurations with multi-turn chat support. See [Appendix](#appendix) for a complete analysis of the workloads.
```

msg:"Before running scorer plugins" pods:[{... "pod_name":"...-5k7qr"}, {... "pod_name":"...-9lp5g"}]
msg:"Pod score" scorer_type:"slo-scorer" pod_name:"vllm-...-9b4wt" score:0.82
msg:"Picked endpoint" scorer_type:"slo-scorer" selected_pod:"vllm-...-9b4wt"
```

---

## Tradeoffs & Gaps

The following areas highlight current limitations and ongoing work for the Predicted-Latency Aware Scheduling system:

- **Addressing Homogeneous Pool Assumptions**
  Current models assume uniform serving pods regarding GPU types and runtimes. Future updates will incorporate richer features and per-variant training to better support heterogeneous pools.

## Takeaway

Accelerator performance is fairly predictable when we account for both the current model server GPU state and request characteristics. By applying online machine learning with a narrow horizon, we can train a model that avoids overfitting while staying accurate to changing workloads. With a good predictor in place, we can route requests based on expected latency, leading to smarter and more efficient load balancing.

---

### Get Involved

> - **Docs & guides**: [https://gateway-api-inference-extension.sigs.k8s.io/guides/latency-based-predictor/](https://gateway-api-inference-extension.sigs.k8s.io/guides/latency-based-predictor/)

Reviewer: same as above, this should just be rendered as a regular hyperlink in markdown:

```suggestion
- **Docs & guides**: https://gateway-api-inference-extension.sigs.k8s.io/guides/latency-based-predictor/
```
